### PR TITLE
UIP-1953 Sync changes from OverReact source

### PR DESCRIPTION
**Pull request by clairesarsam-wf**
## Ultimate problem:

When `OverReact` was split out from its original project, updates continued to be made to the original source at the same time updates were being made to the `OverReact` project.

## How it was fixed:

Merge the updates made in the original project with the updates made in the `OverReact` project.

## Testing suggestions:

* All tests should pass
* Changes made in the original project should now be present in `OverReact`.

## Potential areas of regression:

* Changes made to `OverReact` since it was originally split out should still be present.

---

Please review @greglittlefield-wf @aaronlademann-wf @jacehensley-wf @joelleibow-wf

### CODE REVIEW

**Comment by aviary-wf** [comment]
## Raven
Number of Findings: 0

File: lib/src/component_declaration/component_base.dart [hunk 1]
```diff
@@ -124,12 +131,12 @@ abstract class UiComponent<TProps extends UiProps> extends react.Component {
...
-        if (prop.isNullable && appliedProps.containsKey(prop.key)) return;
-        if (!prop.isNullable && appliedProps[prop.key] != null) return;
+            if (!prop.isRequired) return;
+            if (prop.isNullable && appliedProps.containsKey(prop.key)) return;
+            if (!prop.isNullable && appliedProps[prop.key] != null) return;
 
-        throw new PropError.required(prop.key, prop.errorMessage);
-      });
+            throw new PropError.required(prop.key, prop.errorMessage);
+          });
```
Comment (greglittlefield-wf) [inline lib/src/component_declaration/component_base.dart]: #nit looks like this was fixed in over_react after the code diverged

File: lib/src/component_declaration/component_base.dart [hunk 2]
```diff
@@ -195,11 +203,11 @@ abstract class UiComponent<TProps extends UiProps> extends react.Component {
   // ----------------------------------------------------------------------
 }
 
-/// The basis for a stateful over_react component.
+/// /// The basis for a stateful over_react component.
```
Comment (greglittlefield-wf) [inline lib/src/component_declaration/component_base.dart]: #nit looks like this was fixed in over_react after the code diverged

File: lib/src/component_declaration/component_base.dart [hunk 3]
```diff
@@ -195,11 +203,11 @@ abstract class UiComponent<TProps extends UiProps> extends react.Component {
...
 ///
 /// Includes support for strongly-typed props and state and utilities for prop and CSS classname forwarding.
 ///
-/// Extends [react.Component].
+/// Extends [react.Component]
```
Comment (greglittlefield-wf) [inline lib/src/component_declaration/component_base.dart]: #nit looks like this was fixed in over_react after the code diverged

File: lib/src/component_declaration/component_base.dart [hunk 4]
```diff
@@ -232,7 +240,6 @@ abstract class UiStatefulComponent<TProps extends UiProps, TState extends UiStat
   set unwrappedState(Map value) => super.state = value;
 
   /// Returns a typed state object backed by the specified [stateMap].
-  ///
```
Comment (greglittlefield-wf) [inline lib/src/component_declaration/component_base.dart]: #nit looks like this was fixed in over_react after the code diverged

File: lib/src/component_declaration/flux_component.dart [hunk 5]
```diff
@@ -140,9 +132,9 @@ abstract class _FluxComponentMixin<TProps extends FluxUiProps> implements Batche
     return {};
   }
 
-  /// Register a [subscription] that should be canceled when the component unmounts.
-  ///
-  /// Cancellation will be handled automatically by [componentWillUnmount].
+  /// Register a [subscription] that should be canceled when the component
+  /// unmounts. Cancellation will be handled automatically by
+  /// [componentWillUnmount].
```
Comment (greglittlefield-wf) [inline lib/src/component_declaration/flux_component.dart]: Seems like the version of this file was more up to date in over_react (i.e., these changes should be discarded)

File: lib/src/component_declaration/transformer_helpers.dart [hunk 6]
```diff
@@ -114,47 +126,55 @@ abstract class UiComponent<TProps extends UiProps> extends component_base.UiComp
...
+  @override
+  _RefTypedef get ref => super.ref;
+
+  /// The default consumed prop keys, taken from the keys generated in the associated @[Props] class.
   @toBeGenerated
   Iterable<component_base.ConsumedProps> get $defaultConsumedProps => throw new UngeneratedError(member: #$defaultConsumedProps);
 
   /// The keys for the non-forwarding props defined in this component.
   ///
-  /// For generated components, this defaults to the keys generated in the associated @[annotations.Props] class
```
Comment (greglittlefield-wf) [inline lib/src/component_declaration/transformer_helpers.dart]: #nit looks like this was fixed in over_react after the code diverged

Comment (greglittlefield-wf) [inline lib/src/component_declaration/transformer_helpers.dart]: More of the same in this file

File: lib/src/util/document_event_helper_util.dart [hunk 7]
```diff
@@ -0,0 +1,16 @@
+library document_event_helper;
```
Comment (greglittlefield-wf) [inline lib/src/util/document_event_helper_util.dart]: Not sure this should be added

File: lib/src/component_declaration/component_base.dart [hunk 8]
```diff
@@ -85,14 +86,20 @@ typedef TProps UiFactory<TProps extends UiProps>([Map backingProps]);
...
-///
-/// Related: [UiStatefulComponent]
 abstract class UiComponent<TProps extends UiProps> extends react.Component {
+  /// Returns the component of the specified [ref].
+  /// > `react.Component` if it is a Dart component
+  /// > DOM node if it is a DOM component.
+  ///
+  /// Overridden for strong typing.
+  @override
+  _RefTypedef get ref => super.ref;
```
Comment (jacehensley-wf) [inline lib/src/component_declaration/component_base.dart]: I thought we removed these redeclaration of `ref`s when we added the typing to react-dart.

File: lib/src/util/js_util.dart [hunk 9]
```diff
@@ -0,0 +1,14 @@
+library over_react.js_util;
```
Comment (greglittlefield-wf) [inline lib/src/util/js_util.dart]: Not sure this should be added

File: lib/src/util/rem_util.dart [hunk 10]
```diff
@@ -21,7 +21,6 @@ import 'dart:html';
 import 'package:over_react/over_react.dart';
 import 'package:over_react/src/util/css_value_util.dart';
 import 'package:react/react_dom.dart' as react_dom;
-
```
Comment (greglittlefield-wf) [inline lib/src/util/rem_util.dart]: #Nit there should be a space here

File: lib/src/util/test_mode.dart [hunk 11]
```diff
@@ -13,7 +13,6 @@
 // limitations under the License.
 
 library over_react.test_mode;
-
```
Comment (greglittlefield-wf) [inline lib/src/util/test_mode.dart]: #nit there should be a space here

File: test/over_react/component_declaration/transformer_integration_tests/accessor_mixin_integration_test.dart [hunk 12]
```diff
@@ -50,7 +50,7 @@ main() {
         });
       });
 
-      group('generates prop getters/setters, when there is a custom key namespace, with', () {
+          group('generates prop getters/setters, when there is a custom key namespace, with', () {
```
Comment (greglittlefield-wf) [inline test/over_react/component_declaration/transformer_integration_tests/accessor_mixin_integration_test.dart]: #nit Is this indentation right?

File: lib/src/util/react_wrappers.dart [hunk 13]
```diff
@@ -209,6 +222,12 @@ ReactElement cloneElement(ReactElement element, [Map props, Iterable children])
   }
 }
 
+/// Returns whether the React [instance] is mounted.
+///
+/// Deprecated: Simply call `isMounted` on the [ReactComponent] instead.
+@Deprecated('2.0.0')
+bool isMounted(ReactComponent instance) => instance.isMounted();
```
Comment (jacehensley-wf) [inline lib/src/util/react_wrappers.dart]: I don't think this needs to be added

File: test/over_react/component_declaration/flux_component_test/handler_precedence.dart [hunk 14]
```diff
@@ -20,7 +20,7 @@ class TestHandlerPrecedenceComponent extends FluxUiComponent<TestHandlerPreceden
   @override
   getStoreHandlers() => {props.store.store1: increment};
 
-  increment(Store store) {
```
Comment (jacehensley-wf) [inline test/over_react/component_declaration/flux_component_test/handler_precedence.dart]: I think this had to be typed for strong mode.

File: lib/src/util/document_event_helper_util.dart (hunk 7 above)
Comment (aaronlademann-wf) [inline lib/src/util/document_event_helper_util.dart]: +1 - this remains in ui_core, so there's no reason to have it here.

File: test/over_react/component_declaration/flux_component_test/store_handlers.dart [hunk 15]
```diff
@@ -16,7 +16,7 @@ class TestStoreHandlersComponent extends FluxUiComponent<TestStoreHandlersProps>
   @override
   getStoreHandlers() => {props.store: increment};
 
-  increment(Store store) {
+  increment(_) {
```
Comment (aaronlademann-wf) [inline test/over_react/component_declaration/flux_component_test/store_handlers.dart]: This should not have been changed

File: test/over_react/component_declaration/flux_component_test/store_handlers.dart [hunk 16]
```diff
@@ -7,7 +7,7 @@ UiFactory<TestStoreHandlersProps> TestStoreHandlers;
 class TestStoreHandlersProps extends FluxUiProps<TestActions, TestStore> {}
 
 @Component()
-class TestStoreHandlersComponent extends FluxUiComponent<TestStoreHandlersProps> {
+class TestStoreHandlersComponent extends FluxUiComponent {
```
Comment (aaronlademann-wf) [inline test/over_react/component_declaration/flux_component_test/store_handlers.dart]: This should not have been changed (You should see errors in your analysis server view @clairesarsam-wf  - do you not?)

File: test/over_react/component_declaration/flux_component_test/handler_precedence.dart [hunk 17]
```diff
@@ -7,7 +7,7 @@ UiFactory<TestHandlerPrecedenceProps> TestHandlerPrecedence;
 class TestHandlerPrecedenceProps extends FluxUiProps<TestActions, TestStores> {}
 
 @Component()
-class TestHandlerPrecedenceComponent extends FluxUiComponent<TestHandlerPrecedenceProps> {
+class TestHandlerPrecedenceComponent extends FluxUiComponent {
```
Comment (aaronlademann-wf) [inline test/over_react/component_declaration/flux_component_test/handler_precedence.dart]: This should not have been changed

File: test/over_react/component_declaration/transformer_integration_tests/accessor_mixin_integration_test.dart [hunk 18]
```diff
@@ -81,7 +81,7 @@ main() {
       });
     });
 
-    group('@StateMixin()', () {
+      group('@StateMixin()', () {
```
Comment (aaronlademann-wf) [inline test/over_react/component_declaration/transformer_integration_tests/accessor_mixin_integration_test.dart]: This should not have been changed.

File: test/over_react/component_declaration/transformer_integration_tests/accessor_mixin_integration_test.dart [hunk 19]
```diff
@@ -112,7 +112,7 @@ main() {
         });
       });
 
-      group('generates state getters/setters, when there is a custom key namespace, with', () {
+          group('generates state getters/setters, when there is a custom key namespace, with', () {
```
Comment (aaronlademann-wf) [inline test/over_react/component_declaration/transformer_integration_tests/accessor_mixin_integration_test.dart]: This should not have been changed

File: lib/src/util/document_event_helper_util.dart (hunk 7 above)
Comment (greglittlefield-wf) [inline lib/src/util/document_event_helper_util.dart]: There are things that were part of ui_core that we didn't bring over, and this is one of them.

It's part of the public API maintain, so we shouldn't add it unless there IS a reason to have it here, IMO.

File: test/over_react/util/rem_util_test.dart [hunk 20]
```diff
@@ -167,30 +167,18 @@ main() {
...
 
-      group('throws when passed a CssValue instance with a unit other than px/rem', () {
-        test('', () {
-          expect(() => toPx(new CssValue.parse('1em')), allOf(
-              throwsArgumentError,
-              throwsA(hasToStringValue(contains('must be a rem num or a String px/rem value'))))
-          );
-        });
-
-        test('unless `passThroughUnsupportedUnits` is true', () {
```
Comment (aaronlademann-wf) [inline test/over_react/util/rem_util_test.dart]: I believe this parameter is intentional - shouldn't this test remain?

**Comment by clairesarsam-wf** [comment]
@jacehensley-wf @aaronlademann-wf @greglittlefield-wf Feedback has been addressed - all warnings/infos are gone on dart 1.21.0.

File: test/over_react/component_declaration/component_base_test.dart [hunk 21]
```diff
@@ -92,7 +92,7 @@ main() {
         test('a single child is passed in', () {
           var child = 'Only child';
           var renderedNode = renderAndGetDom(Dom.div()(child));
-          List<Text> children = renderedNode.childNodes.where((node) => node.nodeType != Node.COMMENT_NODE).toList();
+           List<Text> children = renderedNode.childNodes.where((node) => node.nodeType != Node.COMMENT_NODE).toList();
```
Comment (jacehensley-wf) [inline test/over_react/component_declaration/component_base_test.dart]: Indentation is off here.

**Comment by clairesarsam-wf** [comment]
@jacehensley-wf Feedback addressed

**Comment by greglittlefield-wf** [comment]
+1

**Comment by jacehensley-wf** [comment]
+1

**Comment by aaronlademann-wf** [comment]
+1

**Comment by aaronlademann-wf** [comment]
@rmconsole-wf go

**Comment by leviwith-wf** [comment]
QA Resource Approval: +10

- [X] Testing instruction
- [X] Dev +1's
- [X] Dev/QA +10
- [X] Unit test created/updated
- [X] All unit tests pass

Merging

**Pull request merged**